cache/cache_bench.cc

```diff
@@ -220,40 +265,76 @@ class CacheBench {
   }
 
   void OperateCache(ThreadState* thread) {
+    uint64_t result = 0;  // To use looked-up values
+    Cache::Handle* handle = nullptr;
+    KeyGen gen;
     for (uint64_t i = 0; i < FLAGS_ops_per_thread; i++) {
-      uint64_t rand_key = thread->rnd.Next() % FLAGS_max_key;
-      // Cast uint64* to be char*, data would be copied to cache
-      Slice key(reinterpret_cast<char*>(&rand_key), 8);
-      int32_t prob_op = thread->rnd.Uniform(100);
-      if (prob_op >= 0 && prob_op < FLAGS_insert_percent) {
+      Slice key = gen.GetRand(thread->rnd, max_key_);
+      uint32_t prob_op = static_cast<uint32_t>(thread->rnd.Uniform(100U));
+      if (prob_op >= 0 && prob_op < FLAGS_lookup_insert_percent) {
+        if (handle) {
+          cache_->Release(handle);
+          handle = nullptr;
+        }
+        // do lookup
+        handle = cache_->Lookup(key);
+        if (handle) {
+          // do something with the data
+          result += NPHash64(reinterpret_cast<char*>(cache_->Value(handle)),
+                             FLAGS_value_bytes);
+        } else {
+          // do insert
+          cache_->Insert(key, createValue(thread->rnd), FLAGS_value_bytes,
+                         &deleter, &handle);
+        }
+      } else if (prob_op -= FLAGS_lookup_insert_percent,
```

Comment:
`((prob_op -= FLAGS_lookup_insert_percent) < FLAGS_insert_percent)` ? I think the comma operator would trip a lot of people up. Alternatively, we could build a CDF (i.e. precompute `FLAGS_lookup_insert_percent+FLAGS_insert_percent` etc.) so we don't have to subtract.

Reply:
I only gave this enough attention to fix the buggy construction in the existing code (`&&` before `-=`). ;)